### DISCUSSION TITLE
Continuing Stack Overflow Discussion

### DISCUSSION THREAD

**Discussion by kennycarruthers** [trilemma-dev/SecureXPC#107] [General]
Hey Josh,

Happy to follow-up on any questions you have, just be aware that it was a few years ago when I was working on XPC stuff in macOS apps. And I also haven't had a chance to actually use your library, so my comments might be irrelevant if they point out concerns or suggestions that you've already addressed. 

At the time, I was using XPC to do media related things such as: 

- Decoding images (thumbnails, low resolution, high resolution)
- Image editing (Core Image)
- Decoding video frames
- Exporting videos
- File Metadata Harvesting (Extracting EXIF/IPTC info from lots of files.)

Looks like you and Alex are having some good discussions on the library itself. I won't be able to provide much input on that (sorry) but I'm happy to answer any questions about XPC use cases that I was implementing awhile back. 

Original thread from: [Stack Overflow](https://stackoverflow.com/a/72912667/48321)

_Feel free to rename this to something more appropriate._

**Comment by jakaplan** [comment]
Thanks for following up here, I really appreciate it.

A few questions:
- What type of service were you using? XPC service, launch agent, etc?
- What were the reasons to do this out of process? I can imagine stability and untrusted user provided data being amongst them.
- What types of data were you transferring?
- Did the service ever initiate communication with the app, or was it always in response to some sort of request made by the client?
- Only if your comfortable, would you mind mentioning what application this was for?

I completely understand if you don't recall the answers to any or all of these.

**Reply by kennycarruthers** [reply]
> * What type of service were you using? XPC service, launch agent, etc?

XPC Service, embedded inside a traditional AppKit app. 

> * What were the reasons to do this out of process?

Initially it was just for stability. As I got deeper into the topic, I became more interested in how my applications _architecture_ could change (for better or worse) if a lot of the functionality was pushed into XPC services. Ultimately, I took it too far but it was still insightful to play around with moving most functionality into services and having the "app" do nothing more than manage the GUI. 

Security wasn't really a prime concern any more so than the security features I got for free courtesy of the macOS Sandbox. 

> * What types of data were you transferring?

I started with decoding images to display them in a scrolling grid view. The client would send over an `NSURL` and the service would send back an `IOSurface`. Later on, I tried using shared `Metal` textures but struggled at times to get it working properly. 

Things evolved from there when I wanted to an XPC service for some of the use-cases I listed earlier. In general, the client would send over something pretty lightweight, like an `NSURL` or serialized blob of "instructions" and the service would unpack that, process the files and send a response back, generally as an `IOSurface` or as a serialized blob if doing something like harvesting metadata, progress reporting, exporting, etc... 

> * Did the service ever initiate communication with the app?

No. All requests originated in the client, likely as the result of a user-action, and then sent to the service for processing. 

> * Only if your comfortable, would you mind mentioning what application this was for.

Unreleased updates I was working on for two macOS apps I wrote awhile back: Fileloupe and Videloupe
